doc/design/block.md

```diff
@@ -0,0 +1,210 @@
+- help to execute multiple operators, blocks should group operators and runs like a single operator.
+
+## How to use Block
+In `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp`, a sub-block should be used to help to define a sub-block.
+
+Let's start from how a RNNOp is described using Block:
+
+```python
+v = some_op()
+m_boot = some_op()
+
+W = pd.Variable(shape=[20, 20])
+U = pd.Varable(shape=[20, 20])
+
+rnn = create_RNNOp()
+with rnn.stepnet() as net:
+  # declare the input variables that need to be segmented into steps
+  x = net.add_input(v)
+  # declare rnn's memory (state)
+  h = net.add_memory(init=m_boot)
+  
+  fc_out = pd.matmul(W, x)
+  hidden_out = pd.matmul(U, h.pre_memory())
+  sum = pd.add_two(fc_out, hidden_out)
+  act = pd.sigmoid(sum)
+  h.update(act) # update memory
+
+  # declare outputs that needs to be merged across all the steps
+  net.add_output(act, hidden_out)
+
+acts, hs = rnn()
+```
+
+Above Python program builds a protobuf message which describe the model, a C++ Block class will create the corresponding Variables and Operators, and execute all the operators.
+
+## Block Implementation
+
+During the generation of the Protobuf message, the Block should store VarDesc (the Protobuf message which describes Variable) and OpDesc (the Protobuf message which describes Operator).
+
+VarDesc in a block should have its name scope to avoid local variables affect father block's name scope. 
+Child block's name scopes should inherit the father's so that OpDesc in child block can reference a VarDesc that stored in father block. For example
+
+```python
+a = pd.Varaible(shape=[20, 20])
+b = pd.fc(a, params=["fc.w", "fc.b"])
+
+rnn = pd.create_RNNOp()
+with rnn.stepnet() as net:
+    x = net.add_input(a)
+    # reuse fc's parameter
+    fc_without_b = pd.get_variable("fc.w")
+    net.add_output(fc_without_b)
+
+out = rnn()
+```
+the method `pd.get_variable` can help retrieve a Variable by a name, a Variable may store in a father block, but might be retrieved in a child block, so block should have a variable scope that supports inheritance.
+
+In compiler design, the symbol table is an data structure created and maintained by compilers in order to store information about the occurrence of various entities such as variable names, function names, classes, etc.
+
+To store the definition of Variables and Operators, a C++ class `SymbolTable` is introduced as a similar concept with compiler's symbol table.
+
+`SymbolTable` will has following functions:
+
+- store the definitions (some names and attributes) of variables and operators,
+- to verify if a variable name has been declared,
+- to make it possible to implement type checking (offer Protobuf message pointers to `InferShape` handlers).
+
+
+```c++
+// Information in SymbolTable is enough to trace the dependency graph. So maybe
+// the Eval() interface takes a SymbolTable is enough.
+class SymbolTable {
+ public:
+  SymbolTable(SymbolTable* parent) : parent_(parent) {}
```

Comment:
Should `SymbolTable` have `parent` even in compiler?

Reply:
Scope has parent,  SymbolTable stores VarDesc, and VarDesc may exists in SymbolTable's parent.